agent/src/main/java/com/thoughtworks/go/agent/NewRemote.java

```diff
@@ -0,0 +1,235 @@
+                .registerTypeAdapter(ConfigurationProperty.class, new ConfigurationPropertyAdapter())
+                .registerTypeAdapterFactory(builderAdapter())
+                .registerTypeAdapterFactory(materialAdapter())
+                .registerTypeAdapterFactory(workAdapter())
+                .registerTypeAdapterFactory(materialInstanceAdapter())
+                .create();
+    }
+
+    @Override
+    public AgentInstruction ping(AgentRuntimeInfo info) {
+        try {
+            String instruction = this.httpClientHttpInvokerRequestExecutor.doPost("ping", gson.toJson(info));
+            return gson.fromJson(instruction, AgentInstruction.class);
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public Work getWork(AgentRuntimeInfo runtimeInfo) {
+        try {
+            String work = this.httpClientHttpInvokerRequestExecutor.doPost("get_work", gson.toJson(runtimeInfo));
+            return gson.fromJson(work, Work.class);
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public void reportCurrentStatus(AgentRuntimeInfo agentRuntimeInfo, JobIdentifier jobIdentifier, JobState jobState) {
+        Map<String, Object> request = Map.of(
+                "agent_runtime_info", agentRuntimeInfo,
+                "job_identifier", jobIdentifier,
+                "job_state", jobState);
+
+        try {
+            this.httpClientHttpInvokerRequestExecutor.doPost("report_current_status",
+                    gson.toJson(new ReportCurrentStatusRequest(agentRuntimeInfo, jobIdentifier, jobState)));
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public void reportCompleting(AgentRuntimeInfo agentRuntimeInfo, JobIdentifier jobIdentifier, JobResult result) {
+        Map<String, Object> request = Map.of(
+                "agent_runtime_info", agentRuntimeInfo,
+                "job_identifier", jobIdentifier,
+                "job_result", result);
+
+        try {
+            this.httpClientHttpInvokerRequestExecutor.doPost("report_completing",
+                    gson.toJson(new ReportCompleteStatusRequest(agentRuntimeInfo, jobIdentifier, result)));
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public void reportCompleted(AgentRuntimeInfo agentRuntimeInfo, JobIdentifier jobIdentifier, JobResult result) {
+        Map<String, Object> request = Map.of(
+                "agent_runtime_info", agentRuntimeInfo,
+                "job_identifier", jobIdentifier,
+                "job_result", result);
+
+        try {
+            this.httpClientHttpInvokerRequestExecutor.doPost("report_completed",
+                    gson.toJson(new ReportCompleteStatusRequest(agentRuntimeInfo, jobIdentifier, result)));
+        } catch (IOException e) {
```

Comment:
I think the exception needs to be serialised and sent back from the server. It might not always be an IOException. Right?

Reply:
Since this was a quick spike, haven't really thought about error handling. You are right tho, there could be other exceptions, need to take care of it.